File: app/src/main/java/com/adsamcik/signalcollector/Services/TrackerService.java
package com.adsamcik.signalcollector.services;

import android.Manifest;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.location.LocationProvider;
import android.net.wifi.ScanResult;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.os.IBinder;
import android.os.PowerManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;
import android.telephony.TelephonyManager;

import com.adsamcik.signalcollector.utility.Assist;
import com.adsamcik.signalcollector.NoiseTracker;
import com.adsamcik.signalcollector.utility.Preferences;
import com.adsamcik.signalcollector.utility.DataStore;
import com.adsamcik.signalcollector.activities.MainActivity;
import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.data.Data;
import com.adsamcik.signalcollector.interfaces.ICallback;
import com.adsamcik.signalcollector.receivers.NotificationReceiver;
import com.adsamcik.signalcollector.utility.Shortcuts;

import java.lang.ref.WeakReference;
import java.math.RoundingMode;
import java.nio.charset.Charset;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

public class TrackerService extends Service {
	//Constants
	private static final String TAG = "SignalsTracker";
	private final static int LOCK_TIME_IN_MINUTES = 30;
	private final static int LOCK_TIME_IN_MILLISECONDS = LOCK_TIME_IN_MINUTES * Assist.MINUTE_IN_MILLISECONDS;

	private final float MAX_NOISE_TRACKING_SPEED_KM = 18;

	private final long TRACKING_ACTIVE_SINCE = System.currentTimeMillis();


	private static WeakReference<TrackerService> service;

	private static long lockedUntil;
	private static boolean backgroundActivated = false;

	public static ICallback onServiceStateChange;
	public static ICallback onNewDataFound;
	public static Data dataEcho;
	public static int distanceToWifi;

	private Location prevScanPos;
	private long wifiScanTime;

	private final ArrayList<Data> data = new ArrayList<>();
	private LocationListener locationListener;
	private ScanResult[] wifiScanData;
	private LocationManager locationManager;
	private TelephonyManager telephonyManager;
	private WifiManager wifiManager;
	private WifiReceiver wifiReceiver;

	private boolean wifiEnabled = false;

	private int saveAttemptsFailed = 0;

	private NotificationManager notificationManager;
	private PowerManager powerManager;
	private PowerManager.WakeLock wakeLock;

	private NoiseTracker noiseTracker;
	private boolean noiseActive = false;

	/**
	 * Checks if service is running
	 * @return true if service is running
	 */
	public static boolean isRunning() {
		return service != null && service.get() != null;
	}

	/**
	 * Checks if Tracker is auto locked
	 * @return true if locked
	 */
	public static boolean isAutoLocked() {
		return System.currentTimeMillis() < lockedUntil;
	}

	/**
	 * Checks if tracker was activated in background
	 * @return true if activated by the app
	 */
	public static boolean isBackgroundActivated() {
		return backgroundActivated;
	}

	/**
	 * Sets auto lock with predefined time {@link TrackerService#LOCK_TIME_IN_MINUTES}
	 */
	public static void setAutoLock() {
		if (backgroundActivated)
			lockedUntil = System.currentTimeMillis() + LOCK_TIME_IN_MILLISECONDS;
	}

	private void updateData(Location location) {
		if(location.getAltitude() > 5600) {
			stopSelf();
			return;
		}
		wakeLock.acquire();
		Data d = new Data(System.currentTimeMillis());

		if (wifiScanData != null && prevScanPos != null) {
			double timeDiff = (double) (wifiScanTime - prevScanPos.getTime()) / (double) (d.time - prevScanPos.getTime());
			if(timeDiff >= 0) {
				float distTo = location.distanceTo(Assist.interpolateLocation(prevScanPos, location, timeDiff));
				distanceToWifi = (int) distTo;
				//Log.d(TAG, "dist to wifi " + distTo);
				int UPDATE_MAX_DISTANCE_TO_WIFI = 40;
				if (distTo <= UPDATE_MAX_DISTANCE_TO_WIFI && distTo > 0)
					d.setWifi(wifiScanData, wifiScanTime);
			}
			wifiScanData = null;
		}

		SharedPreferences sp = Preferences.get(getApplicationContext());

		if (sp.getBoolean(Preferences.TRACKING_WIFI_ENABLED, true)) {
			wifiManager.startScan();
			prevScanPos = location;
			prevScanPos.setTime(d.time);
		}

		if (sp.getBoolean(Preferences.TRACKING_CELL_ENABLED, true) && !Assist.isAirplaneMode(this)) {
			d.setCell(telephonyManager.getNetworkOperatorName(), telephonyManager.getAllCellInfo());
		}

		if (noiseTracker != null) {
			int evalActivity = Assist.evaluateActivity(ActivityService.lastActivity);
			float MAX_NOISE_TRACKING_SPEED_M = (float) (MAX_NOISE_TRACKING_SPEED_KM / 3.6);
			if ((evalActivity == 1 || (noiseActive && evalActivity == 3)) && location.getSpeed() < MAX_NOISE_TRACKING_SPEED_M) {
				noiseTracker.start();
				short value = noiseTracker.getSample(10);
				if (value >= 0)
					d.setNoise(value);
				noiseActive = true;
			} else {
				noiseTracker.stop();
				noiseActive = false;
			}
		}

		d.setLocation(location).setActivity(ActivityService.lastActivity);

		data.add(d);
		dataEcho = d;

		DataStore.incSizeOfData(DataStore.objectToJSON(d).getBytes(Charset.defaultCharset()).length);

		notificationManager.notify(1, generateNotification(true, d));
		if (onNewDataFound != null)
			onNewDataFound.callback();

		if (data.size() > 10)
			saveData();

		if (backgroundActivated && powerManager.isPowerSaveMode())
			stopSelf();

		wakeLock.release();
	}


	private void saveData() {
		if (data.size() == 0) return;

		SharedPreferences sp = Preferences.get(getApplicationContext());
		Preferences.checkStatsDay(getApplicationContext());

		int wifiCount, cellCount, locations;

		wifiCount = sp.getInt(Preferences.STATS_WIFI_FOUND, 0);
		cellCount = sp.getInt(Preferences.STATS_CELL_FOUND, 0);
		locations = sp.getInt(Preferences.STATS_LOCATIONS_FOUND, 0);
		for (Data d : data) {
			if (d.wifi != null)
				wifiCount += d.wifi.length;
			if (d.cellCount != -1)
				cellCount += d.cellCount;
		}

		sp.edit().putInt(Preferences.STATS_WIFI_FOUND, wifiCount).putInt(Preferences.STATS_CELL_FOUND, cellCount).putInt(Preferences.STATS_LOCATIONS_FOUND, locations + data.size()).apply();

		String input = DataStore.arrayToJSON(data.toArray(new Data[data.size()]));
		input = input.substring(1, input.length() - 1);

		int result = DataStore.saveData(input);
		if (result == 1) {
			saveAttemptsFailed++;
			if (saveAttemptsFailed >= 5)
				stopSelf();
		} else {
			data.clear();
			if (result == 2)
				UploadService.requestUpload(getApplicationContext(), true);
		}
	}

	private Notification generateNotification(boolean gpsAvailable, Data d) {
		Intent intent = new Intent(this, MainActivity.class);
		NotificationCompat.Builder builder = new NotificationCompat.Builder(this)
				.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
				.setSmallIcon(R.drawable.ic_signals_notification)  // the status icon
				.setTicker(getString(R.string.notification_tracker_active_ticker))  // the status text
				.setWhen(System.currentTimeMillis())  // the time stamp
				.setContentIntent(PendingIntent.getActivity(this, 0, intent, 0)) // The intent to send when the entry is clicked
				.setColor(ContextCompat.getColor(getApplicationContext(), R.color.colorPrimary));

		Intent stopIntent = new Intent(this, NotificationReceiver.class);
		stopIntent.putExtra(NotificationReceiver.ACTION_STRING, backgroundActivated ? 0 : 1);
		PendingIntent stop = PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT);
		if (backgroundActivated)
			builder.addAction(R.drawable.ic_battery_alert_black_24dp, getString(R.string.notification_stop_til_recharge), stop);
		else
			builder.addAction(R.drawable.ic_pause, getString(R.string.notification_stop), stop);

		if (!gpsAvailable)
			builder.setContentTitle(getString(R.string.notification_looking_for_gps));
		else {
			builder.setContentTitle(getString(R.string.notification_tracking_active));
			builder.setContentText(buildNotificationText(d));
		}

		return builder.build();
	}

	private String buildNotificationText(final Data d) {
		StringBuilder sb = new StringBuilder();
		DecimalFormat df = new DecimalFormat("#.#");
		df.setRoundingMode(RoundingMode.HALF_UP);
		if (d.wifi != null)
			sb.append(d.wifi.length).append(" wifi ");
		if (d.cellCount != -1)
			sb.append(d.cellCount).append(" cell ");
		if (d.noise > 0)
			sb.append(df.format(Assist.amplitudeToDbm(d.noise))).append(" dB ");
		if (sb.length() > 0)
			sb.setLength(sb.length() - 1);
		else
			sb.append("Nothing found");
		return sb.toString();
	}


	@Override
	public void onCreate() {
		service = new WeakReference<>(this);
		Context appContext = getApplicationContext();
		DataStore.setContext(appContext);

		ActivityService.initializeActivityClient(appContext);

		locationListener = new LocationListener() {
			public void onLocationChanged(Location location) {
				updateData(location);
			}

			public void onStatusChanged(String provider, int status, Bundle extras) {
				if (status == LocationProvider.TEMPORARILY_UNAVAILABLE || status == LocationProvider.OUT_OF_SERVICE)
					notificationManager.notify(1, generateNotification(false, null));
			}

			public void onProviderEnabled(String provider) {
			}

			public void onProviderDisabled(String provider) {
			}
		};

		locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
		telephonyManager = (TelephonyManager) getSystemService(TELEPHONY_SERVICE);
		notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);

		//Enable location update
		int UPDATE_TIME_MILLISEC = 2 * Assist.SECOND_IN_MILLISECONDS;
		float MIN_DISTANCE_M = 5;
		if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
			locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, UPDATE_TIME_MILLISEC, MIN_DISTANCE_M, locationListener);

		wifiManager = (WifiManager) getSystemService(Context.WIFI_SERVICE);

		wifiEnabled = wifiManager.isWifiEnabled();
		if (!wifiEnabled && !wifiManager.isScanAlwaysAvailable())
			wifiManager.setWifiEnabled(true);

		wifiManager.startScan();
		registerReceiver(wifiReceiver = new WifiReceiver(), new IntentFilter(WifiManager.SCAN_RESULTS_AVAILABLE_ACTION));

		powerManager = (PowerManager) getSystemService(POWER_SERVICE);
		wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "TrackerWakeLock");

		if(android.os.Build.VERSION.SDK_INT >= 25 && Shortcuts.initializeShortcuts(this))
			Shortcuts.updateShortcut(this, Shortcuts.TRACKING_ID, getString(R.string.shortcut_stop_tracking), getString(R.string.shortcut_stop_tracking_long), R.drawable.ic_pause, Shortcuts.ShortcutType.STOP_COLLECTION);
	}

	@Override
	public int onStartCommand(Intent intent, int flags, int startId) {
		lockedUntil = 0;
		backgroundActivated = intent == null || intent.getBooleanExtra("backTrack", false);
		startForeground(1, generateNotification(false, null));
		if (onServiceStateChange != null)
			onServiceStateChange.callback();
		if (Preferences.get(this).getBoolean(Preferences.TRACKING_NOISE_ENABLED, false))
			noiseTracker = new NoiseTracker().start();
		return super.onStartCommand(intent, flags, startId);
	}


	@Override
	public void onDestroy() {
		stopForeground(true);
		service = null;

		if (noiseTracker != null)
			noiseTracker.stop();

		if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
			locationManager.removeUpdates(locationListener);
		unregisterReceiver(wifiReceiver);

		saveData();
		if (onServiceStateChange != null)
			onServiceStateChange.callback();
		DataStore.cleanup();

		if (!wifiEnabled)
			wifiManager.setWifiEnabled(false);

		SharedPreferences sp = Preferences.get(getApplicationContext());
		sp.edit().putInt(Preferences.STATS_MINUTES, sp.getInt(Preferences.STATS_MINUTES, 0) + (int) ((System.currentTimeMillis() - TRACKING_ACTIVE_SINCE) / Assist.MINUTE_IN_MILLISECONDS)).apply();

		if(android.os.Build.VERSION.SDK_INT >= 25 && Shortcuts.initializeShortcuts(this))
			Shortcuts.updateShortcut(this, Shortcuts.TRACKING_ID, getString(R.string.shortcut_start_tracking), getString(R.string.shortcut_start_tracking_long), R.drawable.ic_play, Shortcuts.ShortcutType.START_COLLECTION);
	}

	@Nullable
	@Override
	public IBinder onBind(Intent intent) {
		return null;
	}


	private class WifiReceiver extends BroadcastReceiver {
		@Override
		public void onReceive(Context context, Intent intent) {
			wifiScanTime = System.currentTimeMillis();
			List<ScanResult> result = wifiManager.getScanResults();
			wifiScanData = result.toArray(new ScanResult[result.size()]);
		}
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/services/TrackerService.kt
package com.adsamcik.signalcollector.services

import android.Manifest
import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.hardware.SensorManager.SENSOR_DELAY_NORMAL
import android.location.Location
import android.net.wifi.ScanResult
import android.net.wifi.WifiManager
import android.os.Looper
import android.os.PowerManager
import android.telephony.TelephonyManager
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.MutableLiveData
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.activities.LaunchActivity
import com.adsamcik.signalcollector.data.RawData
import com.adsamcik.signalcollector.data.TrackingSession
import com.adsamcik.signalcollector.database.AppDatabase
import com.adsamcik.signalcollector.database.data.DatabaseCellData
import com.adsamcik.signalcollector.database.data.DatabaseWifiData
import com.adsamcik.signalcollector.extensions.getSystemServiceTyped
import com.adsamcik.signalcollector.receivers.NotificationReceiver
import com.adsamcik.signalcollector.utility.*
import com.crashlytics.android.Crashlytics
import com.google.android.gms.location.*
import com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference
import java.math.RoundingMode
import java.text.DecimalFormat

class TrackerService : LifecycleService(), SensorEventListener {
	private var wifiScanTime: Long = 0
	private var wifiScanData: Array<ScanResult>? = null
	private var wifiReceiver: WifiReceiver? = null

	private lateinit var notificationManager: NotificationManager
	private lateinit var powerManager: PowerManager
	private lateinit var wakeLock: PowerManager.WakeLock


	private val telephonyManager: TelephonyManager by lazy { getSystemServiceTyped<TelephonyManager>(Context.TELEPHONY_SERVICE) }
	private val wifiManager: WifiManager by lazy { getSystemServiceTyped<WifiManager>(Context.WIFI_SERVICE) }

	private var minUpdateDelayInSeconds = -1
	private var minDistanceInMeters = -1f

	/**
	 * True if previous collection was mocked
	 */
	private var prevMocked = false

	/**
	 * Previous location of collection
	 */
	private var prevLocation: Location? = null

	private lateinit var locationCallback: LocationCallback

	private val session = TrackingSession(System.currentTimeMillis())

	private var lastStepCount = -1

	private lateinit var keyWifiEnabled: String
	private var defaultWifiEnabled: Boolean = false

	private lateinit var keyCellEnabled: String
	private var defaultCellEnabled: Boolean = false

	private lateinit var keyLocationEnabled: String
	private var defaultLocationEnabled: Boolean = false


	/**
	 * Collects data from necessary places and sensors and creates new RawData instance
	 */
	private fun updateData(location: Location) {
		val distance = if (prevLocation == null) 0f else location.distanceTo(prevLocation)
		if (location.isFromMockProvider) {
			prevMocked = true
			return
		} else if (prevMocked && prevLocation != null) {
			prevMocked = false
			if (distance < minDistanceInMeters)
				return
		}

		wakeLock.acquire(10 * Constants.MINUTE_IN_MILLISECONDS)

		session.apply {
			distanceInM += distance
			collections++
		}

		val preferences = Preferences.getPref(this)
		val d = RawData(System.currentTimeMillis())

		if (preferences.getBoolean(keyWifiEnabled, defaultWifiEnabled)) {
			val prevLocation = prevLocation
			if (prevLocation != null) {
				if (wifiScanData != null) {
					val timeDiff = (wifiScanTime - prevLocation.time).toDouble() / (d.time - prevLocation.time).toDouble()
					if (timeDiff >= 0) {
						val distTo = location.distanceTo(Assist.interpolateLocation(prevLocation, location, timeDiff))
						distanceToWifi = distTo.toInt()

						if (distTo <= UPDATE_MAX_DISTANCE_TO_WIFI && distTo > 0)
							d.setWifi(wifiScanData, wifiScanTime)
					}
					wifiScanData = null
				} else {
					val timeDiff = (wifiScanTime - prevLocation.time).toDouble() / (d.time - prevLocation.time).toDouble()
					if (timeDiff >= 0) {
						val distTo = location.distanceTo(Assist.interpolateLocation(prevLocation, location, timeDiff))
						distanceToWifi += distTo.toInt()
					}
				}
			}

			wifiManager.startScan()
		}

		if (preferences.getBoolean(keyCellEnabled, defaultCellEnabled) && !Assist.isAirplaneModeEnabled(this)) {
			d.addCell(telephonyManager)
		}

		val activityInfo = ActivityService.lastActivity

		if (preferences.getBoolean(keyLocationEnabled, defaultLocationEnabled))
			d.setLocation(location).setActivity(activityInfo)

		rawDataEcho.postValue(d)

		prevLocation = location
		prevLocation!!.time = d.time

		notificationManager.notify(NOTIFICATION_ID_SERVICE, generateNotification(true, d))

		saveData(d)

		if (isBackgroundActivated && powerManager.isPowerSaveMode)
			stopSelf()

		wakeLock.release()
	}

	private fun saveData(data: RawData) {
		GlobalScope.launch {
			val appContext = applicationContext
			val database = AppDatabase.getAppDatabase(appContext)

			database.runInTransaction {
				val location = data.location
				var locationId: Long? = null
				if (location != null)
					locationId = database.locationDao().insert(location.toDatabase(data.activity!!))

				val cell = data.cell
				cell?.registeredCells?.forEach {
					if (database.cellDao().insertWithUpdate(DatabaseCellData(locationId, data.time, data.time, it)) == 0)
						database.cellDao().update(it.cellId, locationId, data.time, it.type, it.asu)
				}

				val wifi = data.wifi
				if (wifi != null) {
					val wifiDao = database.wifiDao()
					wifi.inRange.forEach {
						if (wifiDao.insertWithUpdate(DatabaseWifiData(locationId, wifi.time, wifi.time, it)) == 0)
							wifiDao.update(locationId, it.BSSID, it.SSID, it.capabilities, it.frequency, wifi.time, it.level)
					}
				}
			}
		}
	}

	/**
	 * Generates tracking notification
	 */
	private fun generateNotification(gpsAvailable: Boolean, d: RawData?): Notification {
		val intent = Intent(this, LaunchActivity::class.java)
		val builder = NotificationCompat.Builder(this, getString(R.string.channel_track_id))
				.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
				.setSmallIcon(R.drawable.ic_signals)  // the done icon
				.setTicker(getString(R.string.notification_tracker_active_ticker))  // the done text
				.setWhen(System.currentTimeMillis())  // the time stamp
				.setContentIntent(PendingIntent.getActivity(this, 0, intent, 0)) // The intent to send when the entry is clicked
				.setColor(ContextCompat.getColor(this, R.color.color_accent))

		val stopIntent = Intent(this, NotificationReceiver::class.java)
		stopIntent.putExtra(NotificationReceiver.ACTION_STRING, if (isBackgroundActivated) NotificationReceiver.LOCK_RECHARGE_ACTION else NotificationReceiver.STOP_TRACKING_ACTION)
		val stop = PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT)
		if (isBackgroundActivated) {
			builder.addAction(R.drawable.ic_battery_alert_black_24dp, getString(R.string.notification_stop_til_recharge), stop)

			val stopForMinutes = 60
			val stopForMinutesIntent = Intent(this, NotificationReceiver::class.java)
			stopForMinutesIntent.putExtra(NotificationReceiver.ACTION_STRING, NotificationReceiver.STOP_MINUTES_EXTRA)
			stopForMinutesIntent.putExtra(NotificationReceiver.STOP_MINUTES_EXTRA, stopForMinutes)
			val stopForMinutesAction = PendingIntent.getBroadcast(this, 1, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT)
			builder.addAction(R.drawable.ic_stop_black_24dp, getString(R.string.notification_stop_for_minutes, stopForMinutes), stopForMinutesAction)
		} else
			builder.addAction(R.drawable.ic_pause_circle_filled_black_24dp, getString(R.string.notification_stop), stop)

		if (!gpsAvailable)
			builder.setContentTitle(getString(R.string.notification_looking_for_gps))
		else {
			builder.setContentTitle(getString(R.string.notification_tracking_active))
			builder.setContentText(buildNotificationText(d!!))
		}

		return builder.build()
	}

	/**
	 * Builds text for tracking notification
	 */
	private fun buildNotificationText(d: RawData): String {
		val resources = resources
		val sb = StringBuilder()
		val df = DecimalFormat("#.#")
		df.roundingMode = RoundingMode.HALF_UP

		var isEmpty = true

		if (d.activity != null)
			sb.append(resources.getString(R.string.notification_activity,
					d.activity!!.getResolvedActivityName(this))).append(", ")

		val wifi = d.wifi
		if (wifi != null) {
			sb.append(resources.getString(R.string.notification_wifi, wifi.inRange.size)).append(", ")
			isEmpty = false
		}

		val cell = d.cell
		if (cell != null && cell.registeredCells.isNotEmpty()) {
			val mainCell = cell.registeredCells[0]
			sb.append(resources.getString(R.string.notification_cell_current, mainCell.type.name, mainCell.dbm)).append(' ').append(resources.getQuantityString(R.plurals.notification_cell_count, cell.totalCount, cell.totalCount)).append(", ")
			isEmpty = false
		}

		if (!isEmpty)
			sb.setLength(sb.length - 2)
		else
			sb.append(resources.getString(R.string.notification_nothing_found))

		return sb.toString()
	}


	override fun onCreate() {
		super.onCreate()

		service = WeakReference(this)
		val sp = Preferences.getPref(this)
		val resources = resources

		val packageManager = packageManager

		//Initialize keys and defaults
		keyCellEnabled = resources.getString(R.string.settings_cell_enabled_key)
		keyLocationEnabled = resources.getString(R.string.settings_location_enabled_key)
		keyWifiEnabled = resources.getString(R.string.settings_wifi_enabled_key)

		defaultCellEnabled = resources.getString(R.string.settings_cell_enabled_default)!!.toBoolean()
		defaultLocationEnabled = resources.getString(R.string.settings_location_enabled_default)!!.toBoolean()
		defaultWifiEnabled = resources.getString(R.string.settings_wifi_enabled_default)!!.toBoolean()

		//Get managers
		notificationManager = getSystemServiceTyped(Context.NOTIFICATION_SERVICE)
		powerManager = getSystemServiceTyped(Context.POWER_SERVICE)
		wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "signals:TrackerWakeLock")

		//Database initialization
		GlobalScope.launch {
			val sessionDao = AppDatabase.getAppDatabase(applicationContext).sessionDao()
			session.id = sessionDao.insert(session)
		}

		//Enable location update
		if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
			minUpdateDelayInSeconds = sp.getInt(resources.getString(R.string.settings_tracking_min_time_key), resources.getInteger(R.integer.settings_tracking_min_time_default))
			minDistanceInMeters = sp.getInt(resources.getString(R.string.settings_tracking_min_distance_key), resources.getInteger(R.integer.settings_tracking_min_distance_default)).toFloat()

			val client = LocationServices.getFusedLocationProviderClient(this)
			val request = LocationRequest.create().apply {
				interval = minUpdateDelayInSeconds.toLong() * Constants.SECOND_IN_MILLISECONDS
				fastestInterval = minUpdateDelayInSeconds.toLong()
				smallestDisplacement = minDistanceInMeters
				priority = PRIORITY_HIGH_ACCURACY
			}

			locationCallback = object : LocationCallback() {
				override fun onLocationResult(result: LocationResult) {
					updateData(result.lastLocation)
				}

				override fun onLocationAvailability(availability: LocationAvailability) {
					if (!availability.isLocationAvailable)
						notificationManager.notify(NOTIFICATION_ID_SERVICE, generateNotification(false, null))
				}

			}

			client.requestLocationUpdates(request, locationCallback, Looper.myLooper())
		} else {
			Crashlytics.logException(Exception("Tracker does not have sufficient permissions"))
			stopSelf()
			return
		}

		//Wifi tracking setup
		if (sp.getBoolean(keyWifiEnabled, defaultWifiEnabled)) {
			wifiManager.startScan()
			wifiReceiver = WifiReceiver()
			registerReceiver(wifiReceiver, IntentFilter(WifiManager.SCAN_RESULTS_AVAILABLE_ACTION))
		}

		if(packageManager.hasSystemFeature(PackageManager.FEATURE_SENSOR_STEP_COUNTER)) {
			val sensorManager = getSystemServiceTyped<SensorManager>(Context.SENSOR_SERVICE)
			val stepCounter = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)
			sensorManager.registerListener(this, stepCounter, SENSOR_DELAY_NORMAL)
		}


		//Shortcut setup
		if (android.os.Build.VERSION.SDK_INT >= 25) {
			Shortcuts.initializeShortcuts(this)
			Shortcuts.updateShortcut(this,
					Shortcuts.TRACKING_ID,
					getString(R.string.shortcut_stop_tracking),
					getString(R.string.shortcut_stop_tracking_long),
					R.drawable.ic_pause_circle_filled_black_24dp,
					Shortcuts.ShortcutType.STOP_COLLECTION)
		}
	}

	override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
		super.onStartCommand(intent, flags, startId)

		isBackgroundActivated = intent.getBooleanExtra("backTrack", false)
		startForeground(NOTIFICATION_ID_SERVICE, generateNotification(false, null))

		if (isBackgroundActivated) {
			ActivityService.requestAutoTracking(this, javaClass)
			TrackingLocker.isLocked.observe(this) {
				if (it)
					stopSelf()
			}
		} else
			ActivityService.requestActivity(this, javaClass, minUpdateDelayInSeconds)

		ActivityWatcherService.poke(this, false)

		isServiceRunning.value = true

		return START_NOT_STICKY
	}


	override fun onDestroy() {
		super.onDestroy()

		stopForeground(true)
		service = null
		isServiceRunning.value = false

		ActivityWatcherService.pokeWithCheck(this)
		ActivityService.removeActivityRequest(this, javaClass)

		LocationServices.getFusedLocationProviderClient(this).removeLocationUpdates(locationCallback)

		if (wifiReceiver != null)
			unregisterReceiver(wifiReceiver)

		if (android.os.Build.VERSION.SDK_INT >= 25) {
			Shortcuts.initializeShortcuts(this)
			Shortcuts.updateShortcut(this, Shortcuts.TRACKING_ID, getString(R.string.shortcut_start_tracking), getString(R.string.shortcut_start_tracking_long), R.drawable.ic_play_circle_filled_black_24dp, Shortcuts.ShortcutType.START_COLLECTION)
		}

		val sensorManager = getSystemServiceTyped<SensorManager>(Context.SENSOR_SERVICE)
		sensorManager.unregisterListener(this)


		//Save data to database
		session.end = System.currentTimeMillis()

		GlobalScope.launch {
			val sessionDao = AppDatabase.getAppDatabase(applicationContext).sessionDao()
			sessionDao.update(session)
		}
	}

	override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {

	}

	override fun onSensorChanged(event: SensorEvent) {
		val sensor = event.sensor
		if (sensor.type == Sensor.TYPE_STEP_COUNTER) {
			val stepCount = event.values.first().toInt()
			if (lastStepCount >= 0) {
				//in case sensor would overflow and reset to 0 at some point
				if (lastStepCount > stepCount)
					session.steps += stepCount
				else
					session.steps += stepCount - lastStepCount
			}

			lastStepCount = stepCount
		}
	}

	private inner class WifiReceiver : BroadcastReceiver() {
		override fun onReceive(context: Context, intent: Intent) {
			wifiScanTime = System.currentTimeMillis()
			val result = wifiManager.scanResults
			wifiScanData = result.toTypedArray()
		}
	}

	companion object {
		//Constants
		private const val TAG = "SignalsTracker"
		private const val NOTIFICATION_ID_SERVICE = -7643

		private const val UPDATE_MAX_DISTANCE_TO_WIFI = 40

		/**
		 * LiveData containing information about whether the service is currently running
		 */
		val isServiceRunning = NonNullLiveMutableData(false)

		/**
		 * RawData from previous collection
		 */
		var rawDataEcho = MutableLiveData<RawData>()

		/**
		 * Extra information about distance for tracker
		 */
		var distanceToWifi: Int = 0

		/**
		 * Weak reference to service for AutoLock and check if service is running
		 */
		private var service: WeakReference<TrackerService>? = null

		/**
		 * Checks if tracker was activated in background
		 *
		 * @return true if activated by the app
		 */
		var isBackgroundActivated = false
			private set
	}
}